### PR TITLE
Full state backend

### PR DESCRIPTION
**Pull request by aroradaman**
# FullState Windows Kernel Proxy

## Flow
To program windows proxy we majorly deal in two network components:
1. **HostComputeEndpoints** They represent a virtual port in vSwitch.
2. **HostComputeLoadBalancers** Abstraction which creates VFP rules for Stateful NAT

The idea of fullstate backend is to use a **diffstore** to maintain state (network objects) and track changes. So we create two diffstores:
1. **endpointStore** _for tracking hcn.Endpoints_ 
2. **loadBalancerStore** _for tracking hcn.LoadBalancers_

When we process the fullstate callback, we simply add all the required network objects to achieve the desired network data path in diff stores, after consuming the callback we work on deltas tracked by diffstore and create, update and delete objects accordingly. 

### 1. **hcn/hcn.go** 
Interface for calling host compute network calls and its concrete implementation

### 2. **hcn/fake.go**
Implementation of the interface for testing

### 3. **hcn/endpoint.go**
Endpoint is a user-oriented definition of an HostComputeEndpoint in its entirety.

### 4. **hcn/loadbalancer.go**
LoadBalancer is a user-oriented definition of an HostComputeLoadBalancer in its entirety.

### 5. **hcn/network.go**
Network is a user-oriented definition of an HostComputeNetwork in its entirety.

### 6. **register.go**
Backend registration with KPNG brain

### 7. **store.go**
1. **endpointStore** maintains state for only remote endpoints
2. **loadBalancerStore** maintains state for all load balancers

### 8. **core.go**
1. **PreRun()**  Setting / initializing things
2. **Callback()** KPNG brain calls this callback with a list of ServiceEndpoints (the desired state) which needs to be programmed in the kernel
3. **addServiceEndpointsForClusterIP()**, **addServiceEndpointsForNodePort()** Logic for programming Kubernetes services for Windows kernel, instead of directly programming the network objects we add all the required objects in diffstore
4. **programHostComputeObjects()** This function has side effects, after consuming the callback, this function works on deltas returned by diffstore and programs the kernel accordingly.

### CODE REVIEW

**Comment by jayunit100** [comment]
ok to merge this guy @aroradaman ?

**Comment by aroradaman** [comment]
@jayunit100 yup !
Also, if you are free we can connect tommorow AM to figure out new development environment .. the current AWS EKS environment won't be there for long 😢

**Comment by marosset** [comment]
Please make the OWNERS file updates in a separate PR :)

File: deploy/kpng-windows-capz-calico.yaml [hunk 1]
```diff
@@ -17,14 +17,15 @@ spec:
...
+      serviceAccount: kpng
       serviceAccountName: kpng
       securityContext:
         windowsOptions:
           hostProcess: true
           runAsUserName: "NT AUTHORITY\\system" # TODO: try and run as network service??
       hostNetwork: true
       containers:
-        - image: gcr.io/k8s-staging-win-svc-proxy/kpng:latest
+        - image: damei1807/kpng:v1 #gcr.io/k8s-staging-win-svc-proxy/kpng:latest #
```
Comment (marosset) [inline deploy/kpng-windows-capz-calico.yaml]: Revert this change

File: deploy/kpng-windows-eks.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    spec:
+      serviceAccount: kpng
+      serviceAccountName: kpng
+      securityContext:
+        windowsOptions:
+          hostProcess: true
+          runAsUserName: "NT AUTHORITY\\system" # TODO: try and run as network service??
+      hostNetwork: true
+      containers:
+        - image: damei1807/kpng:v1 #gcr.io/k8s-staging-win-svc-proxy/kpng:latest #
```
Comment (marosset) [inline deploy/kpng-windows-eks.yaml]: gcr.io/k8s-staging-win-svc-proxy/kpng:latest

File: deploy/kpng-windows-capz-calico.yaml [hunk 3]
```diff
@@ -51,7 +52,7 @@ spec:
             # CP node directly since we don't have routes programmed yet.
             # NOTE: These values are specific to the default CAPZ deployment.
             - name: KUBERNETES_SERVICE_HOST
-              value: "10.0.0.4"
+              value: "10.20.30.10"
```
Comment (marosset) [inline deploy/kpng-windows-capz-calico.yaml]: did the default value for CAPZ clusters change here?

**Comment by marosset** [comment]
Overall the change LGTM and thank you for adding the extra documentation!

Please make the OWNERS changes as a separate PR and also revert some of the chances to the deployments so they do not target an individual dockerhub account as mentioned above
After these changes we can merge and get a new image built

/lgtm cancel

File: deploy/kpng-windows-capz-calico.yaml (hunk 3 above)
Comment (aroradaman) [reply deploy/kpng-windows-capz-calico.yaml]: reverting /

**Comment by aroradaman** [comment]
@marosset I have pushed all the required changes

**Review by jayunit100 — APPROVED** [review]
